crates/bevy_hierarchy/src/hierarchy.rs

```diff
@@ -20,6 +23,29 @@ pub struct DespawnChildrenRecursive {
     pub entity: Entity,
 }
 
+/// System to print a warning if at any time we detect an inconsistent hierarchy.
+pub fn hierarchy_healthcheck_system<T: Component>(
+    parent_query: Query<&Parent, (With<T>, Or<(Changed<Parent>, Added<T>)>)>,
+    component_query: Query<(), With<T>>,
+    mut already_diagnosed: Local<bool>,
+) {
+    if *already_diagnosed {
```

Comment:
I understand the desire to avoid doing unnecessary work, but I don't think this is desirable:

1. For apps with "well formed" hierarchies this system will continue to run
2. For apps with "invalid hierarchies", we will only warn the first time for the first scene. If you load another scene (which might also have a bad hierarchy), this print nothing, which will result in a false sense of security.

Seems like the worst of both worlds. If we're going to pay the cost of this system every frame for "good apps", we might as well pay the cost of the system each frame for "bad apps" and warn about every bad entity in a dynamically changing app.

Reply:
I mostly wanted to avoid spamming the log with a message per frame.

If we print once per frame the warning, this forces more or less the user to fix the hierarchy issue as it renders logging unusable, and this becomes as good as panicking.

Seems your concern could be addressed replacing the `bool` by a `HashSet<Entity>` though, I can do that.

I'd be fine with spamming the log if the diagnostic was easily disabled.